aws-java-sdk-core/src/main/java/com/amazonaws/auth/profile/internal/ProfilesConfigFileLoader.java

```diff
@@ -82,43 +85,99 @@
-                                    new BasicSessionCredentials(accessKey, secretKey, sessionToken)));
-            }
+            return new Profile(profileName,
+                    new BasicSessionCredentials(accessKey, secretKey, sessionToken));
         }
+    }
 
-        return profilesByName;
+    private static Profile fromAssumeRole(String profileName, Map<String, String> properties, Map<String, Map<String, String>> allProfileProperties, ProfileCredentialsService profileCredentialsService) {
+        String roleArn = properties.get(Profile.ROLE_ARN);
+        String sourceProfileName = properties.get(Profile.SOURCE_PROFILE);
+        String roleSessionName = properties.get(Profile.ROLE_SESSION_NAME);
+        String externalId = properties.get(Profile.EXTERNAL_ID);
+
+        if (sourceProfileName == null) {
+            throw new AmazonClientException(
+                    String.format("Unable to load credentials into profile [%s]: Source profile name is not specified",
+                            profileName));
+        }
+
+        Map<String, String> sourceProfileProperties = allProfileProperties.get(sourceProfileName);
+
+        if (sourceProfileProperties == null) {
+            throw new AmazonClientException(
+                    String.format("Unable to load source profile [%s]: Source profile not found [%s]",
+                            profileName, sourceProfileName));
+        }
+
+        String sourceAccessKey = sourceProfileProperties.get(Profile.AWS_ACCESS_KEY_ID);
+        String sourceSecretKey = sourceProfileProperties.get(Profile.AWS_SECRET_ACCESS_KEY);
+        String sourceSessionToken = sourceProfileProperties.get(Profile.AWS_SESSION_TOKEN);
+        AWSCredentials sourceCredentials;
+
+        if (sourceAccessKey == null) {
+            throw new AmazonClientException(
+                    String.format("Unable to load credentials into profile [%s]: AWS Access Key ID is not specified in source profile [%s].",
+                            profileName, sourceProfileName));
+        }
+
+        if (sourceSecretKey == null) {
+            throw new AmazonClientException(
+                    String.format("Unable to load credentials into profile [%s]: AWS Secret Access Key is not specified in source profile [%s].",
+                            profileName, sourceProfileName));
+        }
+
+        if (sourceSessionToken == null) {
+            sourceCredentials = new BasicAWSCredentials(sourceAccessKey, sourceSecretKey);
+        } else {
+            sourceCredentials = new BasicSessionCredentials(sourceAccessKey, sourceSecretKey, sourceSessionToken);
+        }
+
+        if (roleSessionName == null) {
+            roleSessionName = "aws-sdk-java-" + System.currentTimeMillis();
+        }
+
+        RoleInfo roleInfo = new RoleInfo(roleArn, roleSessionName, externalId);
+        return new Profile(profileName, sourceProfileName, profileCredentialsService.assumeRole(sourceCredentials, roleInfo), roleInfo);
```

Comment:
It will not attempt to assume the role until the credentials are requested from the Profile object.

Maybe I misunderstood, but previously you said `If we encounter a role based profile and we can't find an appropriate STS implementation then we should fail fast instead of skipping to prevent non deterministic loading of credentials.` I took that to mean that it should fail if any role profile is encountered. I agree that it makes more sense to not fail until the profile credentials are actually requested.